src/nl/hannahsten/texifyidea/psi/LatexCommandsImplUtil.kt

```diff
@@ -218,4 +210,20 @@ fun setName(element: LatexCommands, newName: String): PsiElement {
         element.parent?.node?.replaceChild(oldNode, newNode)
     }
     return element
-}
\ No newline at end of file
+}
+
+fun keyValContentToString(element: LatexKeyvalKey): String =
+    keyValContentToString(element.keyvalContentList)
+
+fun keyValContentToString(list: List<LatexKeyvalContent>): String =
+    list.joinToString(separator = "") {
+        when {
+            it.keyvalText != null -> it.keyvalText!!.text
+            it.parameterGroup != null -> it.parameterGroup!!.parameterGroupText!!.text
```

Comment:
Wait, you are resolving to the _group_ instead of the _text_ in `label={lst:lstinputlisting}`? That explains the find usages behaviour. 

Perhaps I was not entirely clear in my previous comment when I said

> For that, the reference needs to resolve to the lowest level psi element: e.g. lst:inputlisting in \lstinputlisting[label={lst:inputlisting}]{main2.tex}.

So the lowest level psi element is _not_ the group `{lst:inputlisting}` but the text `lst:inputlisting` _inside_ the group.
I think that also saves you some unnecessary gymnastics when renaming, because I see the braces disappear and appear like magic.

Reply:
If I understand your concern correctly: I am actually resolving to `keyval_value` (in `PsiElement.extractLabelElement()`). I had used `parameter_text` initially, but had to change it to `keyval_value` (`getNameIdentifier`, `getName` and `setName`) to support cases like `...,label=some{ label},...`.  That is, the the reference could potentially point to a accumulation of multiple groups/`keyval_text`s. Although I personally wouldn't use something like this, I implemented it for the sake of completeness. The method you are pointing to converts such an accumulation of possibly many different things into a string.

The disappearing braces are indeed an annoyance, but something I couldn't fix (so far). The fact that a simple case like `\lstinputlisting[label={lst:inp<caret>utlisting}]{some/file}` does not work is troubling though and needs further investigation. I assume it is caused by the caret pointing to the inner PSI element and not the `keyval_content`. I think in such a case the inner PSI element would need to resolve to the parent `keyval_value`, meaning that *every* nested component needs to implement the appropriate methods :(.

Reply:
> I assume it is caused by the caret pointing to the inner PSI element and not the keyval_content

Yes, that's how IntelliJ has implemented Find Usages (https://jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/find_usages.html?search=find%20usage) and not something we can change.

>  I think in such a case the inner PSI element would need to resolve to the parent keyval_value

No, the reference needs to resolve _to_ the inner psi element. I can try to do that myself, if you want

Reply:
> ...,label=some{ label},...

I didn't know that was valid LaTeX, it's not making much sense to me at least, so I don't see why we would need to support this.

Reply:
You are right, I just came to the same realization. For some reason I had only tested it with caption (where it seems to work just fine). That simplifies the implementation. To reuse `parameter_text` for that purpose you would have to include the special characters from keyval_text, however (`,label=<blabla>` is valid and seems to work). I guess then you could replace the `commands | keyval_text` part from `keyval_content` with `parameter_text` and remove the `PsiNamed` from `keyval_content`. Do you agree?

Reply:
> I didn't know that was valid LaTeX, it's not making much sense to me at least, so I don't see why we would need to support this. (Basically, my opinion is that TeXiFy should disallow this, because it's too confusing). To support this, apparently the parser needs to avoid constructing a group here in optional parameters, because it's not a group but apparently just two braces.

Disallowing is fine for me. Just out of curiosity: what constitutes a group formally? I was under the impression that groups allow to capture parts of text as is, in which case `...,label=bla{text,text}bla,...` the middle part would need to be a group to escape the comma.

Reply:
Yeah, that sounds like the best solution which is not too much work, I hope you can figure out where all the special characters should go in the parser.

Well... the problem here is that LaTeX was never meant to be parsed as we do, so as long as `{...}` is parsed as a group with children, then `bla{text,text}` can never be a leaf element and thus cannot support Find Usages (besides introducing other inconveniences as you found). What I meant is that I see no reason not to write `{blatext,text}` if you want to escape the comma (which is a common use case, I suppose), at least for labels (different from text in bibtex where you escape capitalization rules, where we don't need it to be a leaf element). We could make an inspection for this later on, to discourage `bla{text,text}` labels, unless we find a good use case for it. Not that I would recommend using commas in labels, but hey, I recently added support for filenames with commas in it as well. It's not your fault LaTeX is so impossible :)